### PR TITLE
Feature/fe be/update motorcycle card

### PR DESCRIPTION
**Pull request by ArtemVolok**
[FE] add opportunity update motorcycle card, add redirect to admin page after success create and update motorcycle card, small fix in logi and style
[BE] add opportunity update motorcycle card, add small fix

### CODE REVIEW

File: backend/src/api/routers/motorcycle.ts [hunk 1]
```diff
@@ -110,11 +141,92 @@ motorcycleRouters.get(
       const motorcycleCard = await MotorcycleCardModel.findById(
         idMotorcycleCard,
       );
+
```
Comment (vladplaz) [inline backend/src/api/routers/motorcycle.ts:124]: move up

File: backend/src/api/routers/motorcycle.ts [hunk 2]
```diff
@@ -89,6 +97,29 @@ motorcycleRouters.delete(
...
       const idMotorcycleCard = req.params.id;
 
+      const motorcycleData = await MotorcycleCardModel.findById(
+        idMotorcycleCard,
+      );
+
+      if (motorcycleData?.id) {
+        const pathToImage = `src/images/${motorcycleData?.uploadImage?.filename}`;
+
+        fs.access(pathToImage, fs.constants.F_OK, (err) => {
```
Comment (vladplaz) [inline backend/src/api/routers/motorcycle.ts]: write a helper function to delete a file by path

File: backend/src/api/routers/motorcycle.ts [hunk 3]
```diff
@@ -89,6 +97,29 @@ motorcycleRouters.delete(
...
+          if (err) {
```
Comment (vladplaz) [inline backend/src/api/routers/motorcycle.ts]: Wrap in promise (better to use fs async methods) and return 500 error if something goes wrong

File: backend/src/api/routers/motorcycle.ts [hunk 4]
```diff
@@ -89,6 +97,29 @@ motorcycleRouters.delete(
     try {
       const idMotorcycleCard = req.params.id;
 
+      const motorcycleData = await MotorcycleCardModel.findById(
+        idMotorcycleCard,
+      );
+
+      if (motorcycleData?.id) {
```
Comment (vladplaz) [inline backend/src/api/routers/motorcycle.ts]: check for uploadImage

File: backend/src/api/routers/motorcycle.ts [hunk 5]
```diff
@@ -110,11 +141,92 @@ motorcycleRouters.get(
...
+motorcycleRouters.put(
+  '/motorcycleCard/:id',
+  authHandler,
+  upload.single('uploadImage'),
+  motorcycleCardValidation,
+  async (req: Request, res: Response, next: NextFunction) => {
+    try {
+      const idMotorcycleCard = req.params.id;
+
+      if (!idMotorcycleCard) {
```
Comment (vladplaz) [inline backend/src/api/routers/motorcycle.ts]: remove this check

File: backend/src/api/routers/motorcycle.ts [hunk 6]
```diff
@@ -110,11 +141,92 @@ motorcycleRouters.get(
...
 );
 
+motorcycleRouters.put(
+  '/motorcycleCard/:id',
+  authHandler,
+  upload.single('uploadImage'),
+  motorcycleCardValidation,
+  async (req: Request, res: Response, next: NextFunction) => {
+    try {
+      const idMotorcycleCard = req.params.id;
```
Comment (vladplaz) [inline backend/src/api/routers/motorcycle.ts]: rename to cardId

File: backend/src/api/routers/motorcycle.ts [hunk 7]
```diff
@@ -110,11 +141,92 @@ motorcycleRouters.get(
...
+      const fileMotorcycle = req.file;
+
+      if (fileMotorcycle) {
+        const motorcycleFromDatabase = await MotorcycleCardModel.findById(
+          idMotorcycleCard,
+        );
+
+        const pathToImage = `src/images/${motorcycleFromDatabase?.uploadImage?.filename}`;
+
+        fs.access(pathToImage, fs.constants.F_OK, (err) => {
```
Comment (vladplaz) [inline backend/src/api/routers/motorcycle.ts]: the same changes as for delete method

File: backend/src/api/routers/motorcycle.ts [hunk 8]
```diff
@@ -110,11 +141,92 @@ motorcycleRouters.get(
...
+            }
+            console.log('File removed successfully');
+          });
+        });
+      }
+
+      const preparedData = fileMotorcycle
+        ? {
+            ...motorcycleCardData,
+            uploadImage: {
```
Comment (vladplaz) [inline backend/src/api/routers/motorcycle.ts:166]: just use uploadImage: fileMotorcycle

File: frontend/src/components/FormCreateMotorcycleCard/index.tsx [hunk 9]
```diff
@@ -0,0 +1,396 @@
...
+import { API_V1_URL } from '../../constants';
+import './style.scss';
+
+interface IFormCreateMotorcycleCard {
+  motorcycleInfo?: IMotorcycleCard | undefined;
+  requestFunction: (data: FormData) => void;
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const FormCreateMotorcycleCard: React.FC<IFormCreateMotorcycleCard> = ({
```
Comment (vladplaz) [inline frontend/src/components/FormCreateMotorcycleCard/index.tsx]: rename to FormMotorcycleCard

File: frontend/src/components/FormCreateMotorcycleCard/index.tsx [hunk 10]
```diff
@@ -0,0 +1,396 @@
...
+    name,
+    numberOfGears,
+    price,
+    typeBrakes,
+    typeCooling,
+    typeMotorcycle,
+    vendorCode,
+    weight,
+    uploadImage: updatedUploadedImage,
+  } = { ...motorcycleInfo };
```
Comment (vladplaz) [inline frontend/src/components/FormCreateMotorcycleCard/index.tsx]: use rest operator, get uploadImage and make other ...rest

File: frontend/src/components/FormCreateMotorcycleCard/index.tsx [hunk 11]
```diff
@@ -0,0 +1,396 @@
...
+
+  const preparedValues = {
```
Comment (vladplaz) [inline frontend/src/components/FormCreateMotorcycleCard/index.tsx]: use spread

File: frontend/src/components/FormCreateMotorcycleCard/index.tsx [hunk 12]
```diff
@@ -0,0 +1,396 @@
...
+    typeCooling,
+    typeMotorcycle,
+    vendorCode,
+    weight,
+  };
+
+  const [useDrag, setUseDrag] = useState<boolean>(false);
+  const [uploadPhoto, setUploadPhoto] = useState<File | null>(null);
+  const [preview, setPreview] = useState<string>('');
+  const [displayUpdatedImage, setDisplayUpdatedImage] = useState<boolean>(
```
Comment (vladplaz) [inline frontend/src/components/FormCreateMotorcycleCard/index.tsx]: use a prop from the outside that tells it's update form

File: frontend/src/components/FormCreateMotorcycleCard/index.tsx [hunk 13]
```diff
@@ -0,0 +1,396 @@
...
+    defaultValues: motorcycleInfo ? preparedValues : defaultValueMotorcycleCard,
+    resolver: yupResolver(createMotorcycleCartSchema),
+  });
+
+  const onSubmit: SubmitHandler<INewMotorcycleCard> = async (
+    data: INewMotorcycleCard,
+  ) => {
+    if (motorcycleInfo && displayUpdatedImage) {
+      // eslint-disable-next-line @typescript-eslint/no-unused-vars
+      const { ['uploadImage']: _, ...preparedData } = data;
```
Comment (vladplaz) [inline frontend/src/components/FormCreateMotorcycleCard/index.tsx]: use just uploadImage

File: frontend/src/components/FormCreateMotorcycleCard/index.tsx [hunk 14]
```diff
@@ -0,0 +1,396 @@
...
+import { IMotorcycleCard } from '../../pages/CatalogMotorcycles/types';
+import { IUpdateMotorcycleCard } from '../../pages/UpdateMotorcycleCard/types';
+
+import DownloadFile from '../../assets/download-solid.svg?react';
+import { API_V1_URL } from '../../constants';
+import './style.scss';
+
+interface IFormCreateMotorcycleCard {
+  motorcycleInfo?: IMotorcycleCard | undefined;
+  requestFunction: (data: FormData) => void;
```
Comment (vladplaz) [inline frontend/src/components/FormCreateMotorcycleCard/index.tsx]: rename to onSubmit and make it promise

File: frontend/src/components/FormCreateMotorcycleCard/index.tsx [hunk 15]
```diff
@@ -0,0 +1,396 @@
...
+        },
+      );
+
+      requestFunction(formData);
+      return;
+    }
+
+    const formData = new FormData();
+
+    Object.entries(data).forEach(([key, value]) => {
```
Comment (vladplaz) [inline frontend/src/components/FormCreateMotorcycleCard/index.tsx]: use a helper function to create a form data from an object

File: frontend/src/components/FormCreateMotorcycleCard/index.tsx [hunk 16]
```diff
@@ -0,0 +1,396 @@
...
+      return setError('uploadImage', {
+        message: 'File should be a photo png type!',
+      });
+    }
+    setUploadPhoto(file);
+    setValue('uploadImage', file);
+  };
+
+  const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    const file = e.target.files;
```
Comment (vladplaz) [inline frontend/src/components/FormCreateMotorcycleCard/index.tsx]: rename to files

File: frontend/src/components/FormCreateMotorcycleCard/index.tsx [hunk 17]
```diff
@@ -0,0 +1,396 @@
...
+
+    const objectUrl = URL.createObjectURL(uploadPhoto);
+    setPreview(objectUrl);
+
+    return () => URL.revokeObjectURL(objectUrl);
+  }, [uploadPhoto]);
+
+  useEffect(() => {
+    if (updatedUploadedImage?.filename && displayUpdatedImage) {
+      const fakeFile = new File(['content'], 'example.jpg', {
```
Comment (vladplaz) [inline frontend/src/components/FormCreateMotorcycleCard/index.tsx]: use name from updatedUploadedImage?.filename

File: frontend/src/components/FormCreateMotorcycleCard/index.tsx [hunk 18]
```diff
@@ -0,0 +1,396 @@
...
+        type: 'image/jpeg',
```
Comment (vladplaz) [inline frontend/src/components/FormCreateMotorcycleCard/index.tsx]: use type from updatedUploadedImage

File: frontend/src/components/FormCreateMotorcycleCard/index.tsx [hunk 19]
```diff
@@ -0,0 +1,396 @@
...
+                {errors[errorName] && (
+                  <p className="form__error">{errors[errorName]?.message}</p>
+                )}
+              </label>
+            </div>
+          );
+        })}
+        <div className="dndBlock">
+          <h2 className="dndBlock-title">Загрузка фото</h2>
+          {!preview && !displayUpdatedImage && (
```
Comment (vladplaz) [inline frontend/src/components/FormCreateMotorcycleCard/index.tsx]: move dndBlock to separate component

File: frontend/src/pages/CatalogMotorcycles/utils.ts [hunk 20]
```diff
@@ -0,0 +1,8 @@
+import axios from 'axios';
```
Comment (vladplaz) [inline frontend/src/pages/CatalogMotorcycles/utils.ts]: move to requests

File: frontend/src/pages/CreateMotorcycleCard/index.tsx [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+import './style.scss';
+import { uploadMotorcycleCardData } from '../../request';
+import { IUpdateMotorcycleCard } from '../UpdateMotorcycleCard/types';
+import { IServerError } from '../../request/types';
+import { ISuccessCreateMotorcycleCard } from './types';
+
+const CreateMotorcycleCardPage = () => {
+  const navigate = useNavigate();
+
+  const { mutate, data } = useMutation<
```
Comment (vladplaz) [inline frontend/src/pages/CreateMotorcycleCard/index.tsx]: add error handling

File: frontend/src/pages/UpdateMotorcycleCard/index.tsx [hunk 22]
```diff
@@ -0,0 +1,92 @@
...
+
+import FormCreateMotorcycleCard from '../../components/FormCreateMotorcycleCard';
+import { getMotorcycleCardById, updateMotorcycleCard } from '../../request';
+import { AdminPageUrl } from '../../UrlsConfig';
+
+import '../CreateMotorcycleCard/style.scss';
+
+const UpdateMotorcycleCard = () => {
+  const navigate = useNavigate();
+  const { id: idFromUrl } = useParams();
```
Comment (vladplaz) [inline frontend/src/pages/UpdateMotorcycleCard/index.tsx]: keep id

File: frontend/src/pages/UpdateMotorcycleCard/index.tsx [hunk 23]
```diff
@@ -0,0 +1,92 @@
...
+  const {
+    data: motorcycleCardData,
+    isLoading,
+    error,
+  } = useQuery<
+    AxiosResponse<{ response: IMotorcycleCard }>,
+    AxiosError<IServerError>
+  >(['motorcycleCard', idFromUrl], {
+    queryFn: () => getMotorcycleCardById(idFromUrl!),
+    refetchOnWindowFocus: false,
```
Comment (vladplaz) [inline frontend/src/pages/UpdateMotorcycleCard/index.tsx]: move to context

File: frontend/src/pages/UpdateMotorcycleCard/index.tsx [hunk 24]
```diff
@@ -0,0 +1,92 @@
...
+    refetchOnMount: 'always',
+    cacheTime: 0,
```
Comment (vladplaz) [inline frontend/src/pages/UpdateMotorcycleCard/index.tsx]: use invalidateQueries

File: frontend/src/types/CreateMotorcycleCards.ts [hunk 25]
```diff
@@ -0,0 +1 @@
+console.log('test console.log');
```
Comment (vladplaz) [inline frontend/src/types/CreateMotorcycleCards.ts]: remove

File: frontend/src/request/types.ts [hunk 26]
```diff
@@ -3,3 +3,5 @@ export interface IServerError {
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
   errors: any;
 }
+
+console.log('test console.log');
```
Comment (vladplaz) [inline frontend/src/request/types.ts]: remove

File: frontend/src/request/index.ts [hunk 27]
```diff
@@ -77,3 +91,5 @@ export const logoutRequest = async () => {
 
   return response;
 };
+
+console.log('test console.log');
```
Comment (vladplaz) [inline frontend/src/request/index.ts]: remove

**Review by vladplaz — CHANGES_REQUESTED** [review]

File: frontend/src/components/FormCreateMotorcycleCard/index.tsx (hunk 18 above)
Comment (ArtemVolok) [reply frontend/src/components/FormCreateMotorcycleCard/index.tsx]: updatedUploadedImage has no type

Comment (vladplaz) [reply frontend/src/components/FormCreateMotorcycleCard/index.tsx]: use type from filename

**Review by vladplaz — APPROVED** [review]

**Pull request merged**